src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/elasticapi/ElasticExtensions.kt

```diff
@@ -43,3 +51,86 @@ fun XContentBuilder.optionalTimeField(name: String, instant: Instant?): XContent
+ */
+suspend fun <C : ElasticsearchClient, T> C.suspendUntil(block: C.(ActionListener<T>) -> Unit): T =
+        suspendCoroutine { cont ->
+            block(object : ActionListener<T> {
+                override fun onResponse(response: T) = cont.resume(response)
+
+                override fun onFailure(e: Exception) = cont.resumeWithException(e)
+            })
+        }
+
+/**
+ * Compares current and previous IndexMetaData to determine if we should create [ManagedIndexConfig].
+ *
+ * If [getPolicyName] returns null then we should not create a [ManagedIndexConfig].
+ * Else if the previous IndexMetaData is null then it means this is a newly created index that should be managed.
+ * Else if the previous IndexMetaData's [getPolicyName] is null then this is an existing index that had
+ * a policy_name added to it.
+ *
+ * @param previousIndexMetaData the previous [IndexMetaData].
+ * @return whether a [ManagedIndexConfig] should be created.
+ */
+fun IndexMetaData.shouldCreateManagedIndexConfig(previousIndexMetaData: IndexMetaData?): Boolean {
+    if (this.getPolicyName() == null) return false
+
+    return previousIndexMetaData?.getPolicyName() == null
+}
+
+/**
+ * Compares current and previous IndexMetaData to determine if we should delete [ManagedIndexConfig].
+ *
+ * If the previous IndexMetaData is null or its [getPolicyName] returns null then there should
+ * be no [ManagedIndexConfig] to delete. Else if the current [getPolicyName] returns null
+ * then it means we should delete the existing [ManagedIndexConfig].
+ *
+ * @param previousIndexMetaData the previous [IndexMetaData].
+ * @return whether a [ManagedIndexConfig] should be deleted.
+ */
+fun IndexMetaData.shouldDeleteManagedIndexConfig(previousIndexMetaData: IndexMetaData?): Boolean {
+    if (previousIndexMetaData?.getPolicyName() == null) return false
+
+    return this.getPolicyName() == null
+}
+
+/**
+ * Compares current and previous IndexMetaData to determine if we should update [ManagedIndexConfig].
+ *
+ * If [getPolicyName] returns null, the previous IndexMetaData does not exist, or the previous IndexMetaData's
+ * [getPolicyName] returns null then we should not update the [ManagedIndexConfig].
+ * Else compare the policy_names and if they are different then we should update.
+ *
+ * @param previousIndexMetaData the previous [IndexMetaData].
+ * @return whether a [ManagedIndexConfig] should be updated.
+ */
+fun IndexMetaData.shouldUpdateManagedIndexConfig(previousIndexMetaData: IndexMetaData?): Boolean {
+    if (this.getPolicyName() == null || previousIndexMetaData?.getPolicyName() == null) return false
+
+    return this.getPolicyName() != previousIndexMetaData.getPolicyName()
+}
+
+/**
+ * Returns the current policy_name if it exists and is valid otherwise returns null.
+ * */
+fun IndexMetaData.getPolicyName(): String? {
+    if (!this.settings.hasValue(ManagedIndexSettings.POLICY_NAME.key)) return null
```

Comment:
nit: I don't see any more benefit doing `hasValue`when we are still going to check for `isNullOrBlank`. You can remove `hasValue` check.